test/gtest/principal-type.cpp

```diff
@@ -0,0 +1,608 @@
+  expectNoCompose(__FILE__, __LINE__, PrincipalType a, PrincipalType b)
+
+TEST(PrincipalTypeTest, ComposeBasic) {
+  Type i32 = Type::i32;
+  Type i64 = Type::i64;
+  Type f32 = Type::f32;
+  Type f64 = Type::f64;
+
+  // Empty types with all combinations of unreachability.
+  // []->[] + []->[] = []->[]
+  COMPOSE(({}, {}), ({}, {}), ({}, {}));
+  // []->[] + []~>[] = []~>[]
+  COMPOSE(({}, {}), ({}, {}, true), ({}, {}, true));
+  // []~>[] + []->[] = []~>[]
+  COMPOSE(({}, {}, true), ({}, {}), ({}, {}, true));
+  // []~>[] + []~>[] = []~>[]
+  COMPOSE(({}, {}, true), ({}, {}, true), ({}, {}, true));
+
+  // i32 in all positions with all combinations of unreachability.
+  // [i32]->[] + []->[] = [i32]->[]
+  COMPOSE(({i32}, {}), ({}, {}), ({i32}, {}));
+  // []->[i32] + []->[] = []->[i32]
+  COMPOSE(({}, {i32}), ({}, {}), ({}, {i32}));
+  // []->[] + [i32]->[] = [i32]->[]
+  COMPOSE(({}, {}), ({i32}, {}), ({i32}, {}));
+  // []->[] + []->[i32] = []->[i32]
+  COMPOSE(({}, {}), ({}, {i32}), ({}, {i32}));
+
+  // [i32]~>[] + []->[] = [i32]~>[]
+  COMPOSE(({i32}, {}, true), ({}, {}), ({i32}, {}, true));
+  // []~>[i32] + []->[] = []~>[i32]
+  COMPOSE(({}, {i32}, true), ({}, {}), ({}, {i32}, true));
+  // []~>[] + [i32]->[] = []~>[]
+  COMPOSE(({}, {}, true), ({i32}, {}), ({}, {}, true));
+  // []~>[] + []->[i32] = []~>[i32]
+  COMPOSE(({}, {}, true), ({}, {i32}), ({}, {i32}, true));
+
+  // [i32]->[] + []~>[] = [i32]~>[]
+  COMPOSE(({i32}, {}), ({}, {}, true), ({i32}, {}, true));
+  // []->[i32] + []~>[] = []~>[]
+  COMPOSE(({}, {i32}), ({}, {}, true), ({}, {}, true));
+  // []->[] + [i32]~>[] = [i32]~>[]
+  COMPOSE(({}, {}), ({i32}, {}, true), ({i32}, {}, true));
+  // []->[] + []~>[i32] = []~>[i32]
+  COMPOSE(({}, {}), ({}, {i32}, true), ({}, {i32}, true));
+
+  // [i32]~>[] + []~>[] = [i32]~>[]
+  COMPOSE(({i32}, {}, true), ({}, {}, true), ({i32}, {}, true));
+  // []~>[i32] + []~>[] = []~>[]
+  COMPOSE(({}, {i32}, true), ({}, {}, true), ({}, {}, true));
+  // []~>[] + [i32]~>[] = []~>[]
+  COMPOSE(({}, {}, true), ({i32}, {}, true), ({}, {}, true));
+  // []~>[] + []~>[i32] = []~>[i32]
+  COMPOSE(({}, {}, true), ({}, {i32}, true), ({}, {i32}, true));
+
+  // A second type (f64) in all possible positions.
+  // [i32]->[i64] + [i64]->[f32] = [i32]->[f32]
+  COMPOSE(({i32}, {i64}), ({i64}, {f32}), ({i32}, {f32}));
+  // [f64 i32]->[i64] + [i64]->[f32] = [f64 i32]->[f32]
+  COMPOSE(({f64, i32}, {i64}), ({i64}, {f32}), ({f64, i32}, {f32}));
+  // [i32 f64]->[i64] + [i64]->[f32] = [f64 i32]->[f32]
+  COMPOSE(({i32, f64}, {i64}), ({i64}, {f32}), ({i32, f64}, {f32}));
+  // [i32]->[f64 i64] + [i64]->[f32] = [i32]->[f64 f32]
+  COMPOSE(({i32}, {f64, i64}), ({i64}, {f32}), ({i32}, {f64, f32}));
+  // [i32]->[i64 f64] + [i64]->[f32] = X
+  NO_COMPOSE(({i32}, {i64, f64}), ({i64}, {f32}));
+  // [i32]->[i64] + [f64 i64]->[f32] = [f64 i32]->[f32]
+  COMPOSE(({i32}, {i64}), ({f64, i64}, {f32}), ({f64, i32}, {f32}));
```

Comment:
I don't follow this one. First we pop an i32 and push an i64. Then we pop an f64 and an i64, that shouldn't work, as an i64 was on the stack?

Reply:
Ah, but the i64 is popped before the f64. So we successfully pop the i64 we had just pushed, then we additionally pop an f64 that must have been there all along.

Reply:
Wait, so `({A, B}, {C, D})` means "pop a B, pop an A; push a C, push a D"? Why isn't the direction of both lists the same?